Firmware/tone04.c

```diff
@@ -41,20 +41,24 @@ void timer4_set_fan0(uint8_t duty)
 	if (duty == 0 || duty == 255)
 	{
 		// We use digital logic if the duty cycle is 0% or 100%
-		TCCR4A &= ~(1 << COM4C1);
-		OCR4C = 0;
+		ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
+		{
+			TCCR4A &= ~_BV(COM4C1);
+			OCR4C = 0;
+		}
 		WRITE(EXTRUDER_0_AUTO_FAN_PIN, duty);
 	}
 	else
 	{
 		// Use the timer for fan speed. Enable the timer compare output and set the duty cycle.
 		// This function also handles the impossible scenario of a fan speed change during a Tone.
 		// Better be safe than sorry.
-		CRITICAL_SECTION_START;
-		// Enable the PWM output on the fan pin.
-		TCCR4A |= (1 << COM4C1);
-		OCR4C = (((uint32_t)duty) * ((uint32_t)((TIMSK4 & (1 << OCIE4A))?OCR4A:255))) / ((uint32_t)255);
-		CRITICAL_SECTION_END;
+		ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
+		{
+			// Enable the PWM output on the fan pin.
+			TCCR4A |= _BV(COM4C1);
+			OCR4C = (((uint32_t)duty) * ((uint32_t)((TIMSK4 & (1 << OCIE4A))?OCR4A:255))) / ((uint32_t)255);
```

Comment:
This is the only part I've seen where the value is not constant, so I would compute the value outside the critical section, then just do the assignment here.

Reply:
Yes, that's a good idea, but I depend on a bit in TIMSK4. I'm afraid of checking it non-atomically, although that could be safe. Please keep in mind that this function is called extremely rarely, so I wouldn't worry that much about the atomic operation delay.

Reply:
is it a problem even if the time between read/write is not atomic? I just worry that this is multiply+divide of a 32bit value, it's quite some cycles even if that's called rarely.

Reply:
ok. will do that.